File: ios/Sinch/SinchCallManager.swift
import AVFoundation
import Foundation
import SinchRTC

@objc public protocol SinchCallManagerDelegate: AnyObject {
  func sinchCallManagerDidStartClient()
  func sinchCallManagerDidFailWithMessage(_ message: String, code: Int)
  func sinchCallManagerRequiresRegistrationCredentials()

  func sinchCallManagerDidReceiveIncomingCall(_ callId: String, remoteUserId: String)
  func sinchCallManagerCallDidProgress(_ callId: String)
  func sinchCallManagerCallDidEstablish(_ callId: String)
  func sinchCallManagerCallDidEnd(_ callId: String, endCause: String)
}

private func endCauseString(_ endCause: SINCallEndCause) -> String {
  switch endCause {
  case .none: return "none"
  case .timeout: return "timeout"
  case .denied: return "denied"
  case .noAnswer: return "noAnswer"
  case .error: return "error"
  case .hungUp: return "hungUp"
  case .canceled: return "canceled"
  case .otherDeviceAnswered: return "otherDeviceAnswered"
  case .inactive: return "inactive"
  case .voIPCallDetected: return "voipCallDetected"
  case .gsmCallDetected: return "gsmCallDetected"
  @unknown default: return "none"
  }
}

@objc(SinchCallManager)
public final class SinchCallManager: NSObject {
  @objc public weak var delegate: SinchCallManagerDelegate?

  private var client: SINClient?
  private var pendingRegistration: SINClientRegistration?
  private var calls: [String: SINCall] = [:]

  @objc public func configure(appKey: String, environmentHost: String, userId: String) {
    do {
      let sinchClient = try Sinch.client(
        withApplicationKey: appKey,
        environmentHost: environmentHost,
        userId: userId
      )
      sinchClient.delegate = self
      sinchClient.call().delegate = self
      client = sinchClient
    } catch {
      let nsError = error as NSError
      delegate?.sinchCallManagerDidFailWithMessage(nsError.localizedDescription, code: nsError.code)
    }
  }

  @objc public func start() {
    client?.start()
  }

  @objc public func stop() {
    client?.terminateGracefully()
    client = nil
    calls.removeAll()
  }

  @objc public func provideRegistrationCredentials(_ jwt: String) {
    pendingRegistration?.register(withJWT: jwt)
    pendingRegistration = nil
  }

  @objc public func failRegistration(_ message: String) {
    let error = NSError(
      domain: "SinchCalling",
      code: -1,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
    pendingRegistration?.registerDidFail(error)
    pendingRegistration = nil
  }

  @objc public func callUser(_ userId: String) -> String {
    guard let call = client?.call().callUser(withId: userId) else {
      return ""
    }
    track(call)
    return call.callId
  }

  @objc public func callPhoneNumber(_ phoneNumber: String) -> String {
    guard let call = client?.call().callPhoneNumber(phoneNumber) else {
      return ""
    }
    track(call)
    return call.callId
  }

  @objc public func callConference(_ conferenceId: String) -> String {
    guard let call = client?.call().callConference(withId: conferenceId) else {
      return ""
    }
    track(call)
    return call.callId
  }

  @objc public func answerCall(_ callId: String) {
    calls[callId]?.answer()
  }

  @objc public func hangupCall(_ callId: String) {
    calls[callId]?.hangup()
  }

  @objc public func sendDTMF(_ callId: String, key: String) -> Bool {
    return calls[callId]?.sendDTMF(key) ?? false
  }

  @objc public func setMuted(_ muted: Bool) {
    if muted {
      client?.audioController().mute()
    } else {
      client?.audioController().unmute()
    }
  }

  @objc public func setSpeakerEnabled(_ enabled: Bool) {
    if enabled {
      client?.audioController().enableSpeaker()
    } else {
      client?.audioController().disableSpeaker()
    }
  }

  @objc public func enableManagedPushNotifications() {
    client?.enableManagedPushNotifications()
  }

  @objc public func relayPushNotification(_ payload: [AnyHashable: Any]) {
    _ = client?.relayPushNotification(payload)
  }

  @objc public func didActivateAudioSession(_ audioSession: AVAudioSession) {
    client?.call().didActivate(audioSession)
  }

  @objc public func didDeactivateAudioSession(_ audioSession: AVAudioSession) {
    client?.call().didDeactivate(audioSession)
  }

  private func track(_ call: SINCall) {
    calls[call.callId] = call
    call.delegate = self
  }
}

extension SinchCallManager: SINClientDelegate {
  public func clientDidStart(_ client: SINClient) {
    delegate?.sinchCallManagerDidStartClient()
  }

  public func clientDidFail(_ client: SINClient, error: Error) {
    let nsError = error as NSError
    delegate?.sinchCallManagerDidFailWithMessage(nsError.localizedDescription, code: nsError.code)
  }

  public func client(
    _ client: SINClient,
    requiresRegistrationCredentials registrationCallback: SINClientRegistration
  ) {
    pendingRegistration = registrationCallback
    delegate?.sinchCallManagerRequiresRegistrationCredentials()
  }
}

extension SinchCallManager: SINCallClientDelegate {
  public func client(_ client: SINCallClient, didReceiveIncomingCall call: SINCall) {
    track(call)
    delegate?.sinchCallManagerDidReceiveIncomingCall(call.callId, remoteUserId: call.remoteUserId)
  }
}

extension SinchCallManager: SINCallDelegate {
  public func callDidProgress(_ call: SINCall) {
    delegate?.sinchCallManagerCallDidProgress(call.callId)
  }

  public func callDidEstablish(_ call: SINCall) {
    delegate?.sinchCallManagerCallDidEstablish(call.callId)
  }

  public func callDidEnd(_ call: SINCall) {
    calls.removeValue(forKey: call.callId)
    delegate?.sinchCallManagerCallDidEnd(call.callId, endCause: endCauseString(call.details.endCause))
  }
}

File: android/src/main/java/com/sinchcalling/SinchCallManager.kt
package com.sinchcalling

import android.content.Context
import android.os.PowerManager
import com.sinch.android.rtc.AudioController
import com.sinch.android.rtc.ClientRegistration
import com.sinch.android.rtc.PushConfiguration
import com.sinch.android.rtc.SinchClient
import com.sinch.android.rtc.SinchClientListener
import com.sinch.android.rtc.SinchError
import com.sinch.android.rtc.SinchPush
import com.sinch.android.rtc.calling.Call
import com.sinch.android.rtc.calling.CallController
import com.sinch.android.rtc.calling.CallControllerListener
import com.sinch.android.rtc.calling.CallEndCause
import com.sinch.android.rtc.calling.CallListener
import com.sinch.android.rtc.calling.MediaConstraints
import com.sinchcalling.telecom.SinchTelecomListener
import com.sinchcalling.telecom.SinchTelecomManager
import java.io.IOException

interface SinchCallManagerListener {
  fun onClientStarted()
  fun onClientFailed(message: String, code: Int)
  fun onRegistrationCredentialsRequired()

  fun onIncomingCall(callId: String, remoteUserId: String)
  fun onCallProgressing(callId: String)
  fun onCallEstablished(callId: String)
  fun onCallEnded(callId: String, endCause: String)

  fun onPushTokenRegistered()
  fun onPushTokenRegistrationFailed(message: String, code: Int)

  // A system call UI was reported — either the consumer app called
  // `reportIncomingCallUI` directly, or a configured custom push was
  // auto-detected (see `configureCustomIncomingCallPush`). Distinct from a
  // Sinch-relayed push (e.g. an app-to-app incoming call), which instead
  // surfaces as `onIncomingCall`.
  fun onIncomingCallUIShown(callId: String, displayName: String)
  fun onCallUIAnswered(callId: String)
  fun onCallUIDeclined(callId: String)
}

private fun endCauseString(endCause: CallEndCause): String = when (endCause) {
  CallEndCause.NONE -> "none"
  CallEndCause.TIMEOUT -> "timeout"
  CallEndCause.DENIED -> "denied"
  CallEndCause.NO_ANSWER -> "noAnswer"
  CallEndCause.FAILURE -> "error"
  CallEndCause.HUNG_UP -> "hungUp"
  CallEndCause.CANCELED -> "canceled"
  CallEndCause.OTHER_DEVICE_ANSWERED -> "otherDeviceAnswered"
  CallEndCause.INACTIVE -> "inactive"
  CallEndCause.VOIP_CALL_DETECTED -> "voipCallDetected"
  CallEndCause.GSM_CALL_DETECTED -> "gsmCallDetected"
  else -> "none"
}

class SinchCallManager(private val context: Context) {
  var listener: SinchCallManagerListener? = null

  private var client: SinchClient? = null
  private var pendingRegistration: ClientRegistration? = null
  private var pendingPushConfiguration: PushConfiguration? = null
  private val calls = mutableMapOf<String, Call>()
  private var customPushIdField: String? = null
  private var customPushDisplayField: String? = null

  // Proximity-screen-off during an established earpiece call, same as the
  // system Phone app: on while at least one call is established and the
  // speaker isn't in use, off otherwise.
  private val establishedCallIds = mutableSetOf<String>()
  private var isSpeakerEnabled = false
  private var proximityWakeLock: PowerManager.WakeLock? = null

  private val callListener = object : CallListener {
    override fun onCallProgressing(call: Call) {
      listener?.onCallProgressing(call.callId)
    }

    override fun onCallEstablished(call: Call) {
      SinchTelecomManager.reportCallActive(call.callId)
      establishedCallIds.add(call.callId)
      updateProximityMonitoring()
      listener?.onCallEstablished(call.callId)
    }

    override fun onCallEnded(call: Call) {
      calls.remove(call.callId)
      establishedCallIds.remove(call.callId)
      updateProximityMonitoring()
      SinchTelecomManager.reportCallEnded(call.callId)
      listener?.onCallEnded(call.callId, endCauseString(call.details.endCause))
    }
  }

  init {
    SinchTelecomManager.registerPhoneAccount(context.applicationContext)
    SinchTelecomManager.listener = object : SinchTelecomListener {
      override fun onAnswer(callId: String) {
        answerCall(callId)
      }

      override fun onReject(callId: String) {
        hangupCall(callId)
      }

      override fun onDisconnect(callId: String) {
        hangupCall(callId)
      }

      override fun onAnswerExternal(callId: String) {
        listener?.onCallUIAnswered(callId)
      }

      override fun onDeclineExternal(callId: String) {
        listener?.onCallUIDeclined(callId)
      }
    }
  }

  fun registerFcmPush(senderId: String, token: String) {
    pendingPushConfiguration = PushConfiguration.fcmPushConfigurationBuilder()
      .senderID(senderId)
      .registrationToken(token)
      .build()
  }

  // Opts into recognizing the consumer app's own backend's incoming-call
  // push payloads. Until called, every payload is relayed straight to
  // Sinch (see `relayRemotePushNotification`).
  fun configureCustomIncomingCallPush(idField: String, displayField: String) {
    customPushIdField = idField
    customPushDisplayField = displayField
  }

  fun relayRemotePushNotification(payload: Map<String, String>) {
    val idField = customPushIdField
    val callId = idField?.let { payload[it] }
    if (callId != null) {
      val displayName = customPushDisplayField?.let { payload[it] } ?: ""
      reportIncomingCallUI(callId, displayName)
      return
    }

    val result = SinchPush.queryPushNotificationPayload(context.applicationContext, payload)
    client?.relayRemotePushNotification(result)
  }

  // Reports a system call UI (CallKit / Telecom) for a call the consumer
  // app decided to show — from a custom push, a Socket.IO event, or
  // anything else. Independent of `configureCustomIncomingCallPush`.
  fun reportIncomingCallUI(callId: String, displayName: String) {
    SinchTelecomManager.reportExternalIncomingCall(context.applicationContext, callId, displayName)
    listener?.onIncomingCallUIShown(callId, displayName)
  }

  fun configure(appKey: String, environmentHost: String, userId: String) {
    val builder = SinchClient.builder()
      .context(context.applicationContext)
      .applicationKey(appKey)
      .environmentHost(environmentHost)
      .userId(userId)

    pendingPushConfiguration?.let { builder.pushConfiguration(it) }

    val sinchClient = try {
      builder.build()
    } catch (error: IOException) {
      listener?.onClientFailed(error.message ?: "Failed to build Sinch client", -1)
      return
    }

    sinchClient.audioController.enableAutomaticAudioRouting(
      AudioController.AudioRoutingConfig(AudioController.UseSpeakerphone.SPEAKERPHONE_AUTO, true)
    )

    sinchClient.addSinchClientListener(object : SinchClientListener {
      override fun onClientStarted(client: SinchClient) {
        listener?.onClientStarted()
      }

      override fun onClientFailed(client: SinchClient, error: SinchError) {
        listener?.onClientFailed(error.message ?: "Unknown Sinch client error", error.code)
      }

      override fun onCredentialsRequired(clientRegistration: ClientRegistration) {
        pendingRegistration = clientRegistration
        listener?.onRegistrationCredentialsRequired()
      }

      override fun onPushTokenRegistered() {
        listener?.onPushTokenRegistered()
      }

      override fun onPushTokenRegistrationFailed(error: SinchError) {
        listener?.onPushTokenRegistrationFailed(error.message ?: "Unknown Sinch push error", error.code)
      }

      override fun onPushTokenUnregistered() {}
      override fun onPushTokenUnregistrationFailed(error: SinchError) {}

      override fun onUserRegistered() {}
      override fun onUserRegistrationFailed(error: SinchError) {}
    })

    sinchClient.callController.addCallControllerListener(object : CallControllerListener {
      override fun onIncomingCall(callController: CallController, call: Call) {
        track(call)
        SinchTelecomManager.reportIncomingCall(
          context.applicationContext,
          call.callId,
          call.remoteUserId
        )
        listener?.onIncomingCall(call.callId, call.remoteUserId)
      }
    })

    client = sinchClient
  }

  fun start() {
    client?.start()
  }

  fun stop() {
    client?.terminateGracefully()
    client = null
    calls.clear()
    establishedCallIds.clear()
    isSpeakerEnabled = false
    releaseProximityWakeLock()
  }

  fun provideRegistrationCredentials(jwt: String) {
    pendingRegistration?.register(jwt)
    pendingRegistration = null
  }

  fun failRegistration() {
    pendingRegistration?.registerFailed()
    pendingRegistration = null
  }

  fun callUser(userId: String): String {
    val call = client?.callController?.callUser(userId, MediaConstraints(false)) ?: return ""
    track(call)
    return call.callId
  }

  fun callPhoneNumber(phoneNumber: String, callerId: String): String {
    val call = client?.callController?.callPhoneNumber(phoneNumber, callerId) ?: return ""
    track(call)
    return call.callId
  }

  fun callConference(conferenceId: String, callerId: String): String {
    val call = client?.callController?.callConference(conferenceId, callerId) ?: return ""
    track(call)
    return call.callId
  }

  // Turns a call shown via `reportIncomingCallUI` into a real call once you
  // know which Sinch conference to join: joins the conference and
  // re-associates the already-visible system call UI with the resulting
  // real callId. Returns the real callId, or "" on failure.
  fun resolveCallUIToConference(callId: String, conferenceId: String, callerId: String): String {
    val call = client?.callController?.callConference(conferenceId, callerId)
    if (call == null) {
      SinchTelecomManager.reportCallEnded(callId)
      return ""
    }
    track(call)
    SinchTelecomManager.remapCallId(callId, call.callId)
    return call.callId
  }

  // Cleans up a system call UI reported via `reportIncomingCallUI` that
  // couldn't be resolved into a real call.
  fun dismissCallUI(callId: String) {
    SinchTelecomManager.reportCallEnded(callId)
  }

  fun updateIncomingCallDisplayName(callId: String, displayName: String) {
    SinchTelecomManager.updateCallerDisplayName(callId, displayName)
  }

  fun answerCall(callId: String) {
    calls[callId]?.answer()
  }

  fun hangupCall(callId: String) {
    calls[callId]?.hangup()
  }

  fun sendDTMF(callId: String, key: String): Boolean {
    val call = calls[callId] ?: return false
    call.sendDTMF(key)
    return true
  }

  fun setMuted(muted: Boolean) {
    val audioController = client?.audioController ?: return
    if (muted) audioController.mute() else audioController.unmute()
  }

  fun setSpeakerEnabled(enabled: Boolean) {
    val audioController = client?.audioController ?: return
    if (enabled) audioController.enableSpeaker() else audioController.disableSpeaker()
    isSpeakerEnabled = enabled
    updateProximityMonitoring()
  }

  private fun updateProximityMonitoring() {
    val shouldEnable = establishedCallIds.isNotEmpty() && !isSpeakerEnabled
    if (shouldEnable) acquireProximityWakeLock() else releaseProximityWakeLock()
  }

  private fun acquireProximityWakeLock() {
    val lock = proximityWakeLock ?: run {
      val powerManager =
        context.applicationContext.getSystemService(Context.POWER_SERVICE) as? PowerManager
          ?: return
      if (!powerManager.isWakeLockLevelSupported(PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK)) {
        return
      }
      @Suppress("DEPRECATION")
      powerManager.newWakeLock(
        PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK,
        "SinchCalling:ProximityWakeLock",
      ).also { proximityWakeLock = it }
    }
    if (!lock.isHeld) {
      lock.acquire(10 * 60 * 1000L)
    }
  }

  private fun releaseProximityWakeLock() {
    proximityWakeLock?.let { if (it.isHeld) it.release() }
  }

  private fun track(call: Call) {
    calls[call.callId] = call
    call.addCallListener(callListener)
  }
}
